lib/channel-postmessage/src/index.ts

```diff
@@ -55,27 +55,43 @@ export class PostmsgTransport {
    * @param event
    */
   send(event: ChannelEvent, options?: any): Promise<any> {
-    const iframeWindow = this.getWindow();
-    if (!iframeWindow || this.buffer.length) {
-      return new Promise((resolve, reject) => {
-        this.buffer.push({ event, resolve, reject });
-      });
-    }
     let depth = 15;
     let allowFunction = true;
+    let target;
 
     if (options && typeof options.allowFunction === 'boolean') {
       allowFunction = options.allowFunction;
     }
     if (options && Number.isInteger(options.depth)) {
       depth = options.depth;
     }
+    if (options && typeof options.target === 'string') {
+      target = options.target;
+    }
+
+    const frames = this.getFrames(target);
 
-    const data = stringify({ key: KEY, event }, { maxDepth: depth, allowFunction });
+    const data = stringify(
+      { key: KEY, event, source: document.location.origin + document.location.pathname },
+      { maxDepth: depth, allowFunction }
+    );
 
     // TODO: investigate http://blog.teamtreehouse.com/cross-domain-messaging-with-postmessage
     // might replace '*' with document.location ?
-    iframeWindow.postMessage(data, '*');
+    frames.forEach(f => {
+      try {
+        f.postMessage(data, '*');
+      } catch (e) {
+        console.error('sending over postmessage fail');
+      }
+    });
+
+    if (!frames.length || this.buffer.length) {
+      return new Promise((resolve, reject) => {
+        this.buffer.push({ event, resolve, reject });
+      });
+    }
```

Comment:
Does this mean we buffer messages until the first frame appears, and then send them all to just that first frame?

We probably need to be careful that rather than the *first* frame, the **local** frame is the one we (a) wait for and (b) send the buffered messages to.

Reply:
I changed it so it's guaranteed that the local/main frame will get the flush.